### PR TITLE
Added Party Characteristics Tab and Visualizations to Dashboard

### LINKED ISSUE #249: Party Characteristics visualizations not showing up for projects outside West Cascades
The `SiteID` column in [party_characteristics.csv](https://github.com/OutdoorRD/trails-viz-data/blob/master/chatbot/party_characteristics.csv) contains Strings. A manual conversion from String to List is necessary to ensure the `SiteID` column has data of type List and not type String. This allows `trails-viz-api` endpoints to perform a successful iteration across the sites within the list.

### PR DESCRIPTION
**Pull request by davidye007**
Implemented both API endpoints and UI for the Party Characteristics Tab:
1. Number of Responses (Table)
2. Yearly Average Response (Timeseries)
3. Distribution of Responses (Side-by-side Bar Chart)

**Note:** As of now, the Party Characteristics tab will show up for **all available projects** (assuming user has logged in and access to the Visitor Characteristics tab). In other words, any project that includes sites with chatbot data will have party characteristic visualizations available.

Currently, there isn't a way to toggle which projects have access to which tabs (e.g. Party Characteristics, Home Locations, Demographics). Though this could be a useful functionality for us to add.

Fixes #249 
Fixes OutdoorRD/west-cascades#159

### CODE REVIEW

**Comment by EmiliaH** [comment]
Thanks @davidye007! Cool to see the PR for this! @jblehr, curious to know if you have any ideas on a simple way to allow us to specify which projects display this data? As David mentioned currently any site with chatbot data will show these data visualizations.

**Comment by EmiliaH** [comment]
Also adding here that we talked about a couple additional improvements that we're leaving on the table for now. Specifically highlighting this one:

> Convert number of responses table into a bar chart with x-axis as years, y-axis as counts, and categories as the total number of conversations, as well as each of the 4 party characteristics. This may also eliminate the need of the count display mode in showing distribution of responses.

From [this comment](https://github.com/OutdoorRD/west-cascades/issues/159#issuecomment-2521220326) on [west-cascades/issue159 ](https://github.com/orgs/OutdoorRD/projects/38/views/1?pane=issue&itemId=80293687&issue=OutdoorRD%7Cwest-cascades%7C159)

**Comment by jblehr** [comment]
Nice work @davidye007! One request before I get started on this: Can you please open (and link) the corresponding PR in `trails-viz-data` with the data I should use to test the new functionality?

That way, I can run this in dev mode on my computer to test it out. Or, is there a newer/better way for me to do test/review this?

**Comment by davidye007** [comment]
@jblehr the chatbot data in `trails-viz-data` is already on the master branch found [here](https://github.com/OutdoorRD/trails-viz-data/tree/master/chatbot). I believe running `git pull` inside the `trails-viz-data` repo will be the quickest way to get the data locally and test on your computer.

I locally tested the `Party Characteristics` tab a bit with clicking around on different sites as well as projects. I think the main thing would be for you to check if the tab and visualizations run smoothly on your computer as well (e.g. maybe click on a few different sites/projects).

**Comment by davidye007** [comment]
closes #259

**Comment by davidye007** [comment]
**TL;DR**: The Party Characteristics tab renders plots for `PartyPeople`, `PartyVehics`, `PeoplePerVehics`, and `TrailVisits`. Missing any of these features in `chatbot_data.csv`? No problem—only the plots for available features will be displayed!

---

The `chatbot_data` endpoints retrieve data from `chatbot_data.csv`, which is generated by the `chatbot_data_prep` Makefile recipe in the `vic-flaskbot` repository. This recipe allows users to specify which features to include in `chatbot_data.csv`. Previously, if the user excluded any required features (`PartyPeople`, `PartyVehics`, or `TrailVisits`) while running the recipe, the visualizations relying on `chatbot_data.csv` would fail to load. This occurred because a single endpoint handled data for all `PartyCharacteristics` features; if any of the required features were missing, the API would crash, leaving the Party Characteristics tab empty.

To resolve this issue, I refactored the API to handle requests for individual features (e.g., `PartyPeople`, `PartyVehics`, `TrailVisits`, or `PeoplePerVehics`). The frontend now makes separate API calls for each feature and conditionally renders plots based on the returned status code. If a feature is missing from `chatbot_data.csv`, its endpoint responds with a `204 No Content` status code, and the corresponding plot is omitted. This ensures that visualizations for all available features are still displayed, even if some features are missing.

Additionally, I renamed the endpoints from `PartyCharacteristics` to `ChatbotData`, as the endpoints are now more generalizable, and can easily accommodate other features that we choose to include in `chatbot_data.csv` (e.g. ZipCode, InfoSource, Distance, etc.).

**Comment by davidye007** [comment]
@jblehr I opened a [PR ](https://github.com/OutdoorRD/vic-flaskbot/pull/241) for the `make` target that generates `chatbot_data.csv`. With the make target you should be able to generate the data needed to test out the `Party Characteristics` tab. Let me know if you run into any issues or have any questions/concerns. Thanks a lot!

P.S. It might be good to pull from the `PartyCharacteristics` branch again, since I made quite a bit of refactoring changes since initially opening this PR (see the latest update in the comment above).

**Comment by davidye007** [comment]
Hey @jblehr, I just finished linting the code and the PR successfully passed the CI/CD checkpoints. Could you please check if the dashboard runs locally on your computer? 

To generate the chatbot data used by the dashboard, you can either:
1. Generate the data using the [chatbot data prep make target](https://github.com/OutdoorRD/vic-flaskbot/pull/241). The steps are outlined [here](https://github.com/OutdoorRD/vic-flaskbot/wiki/Parsing-and-munging-survey-responses#prepare-chatbot-data-for-trailtrends-api).
2. Pull the data directly from [trails-viz-data PR #74](https://github.com/OutdoorRD/trails-viz-data/pull/74). This is the data generated from `1.` and will be merged into `main` (deployed on the `droplet`) simultaneously with this PR.

Please let me know if you have any questions or concerns, thanks!

**Comment by davidye007** [comment]
Documenting issues this PR covers.

`trails-viz`: Fixes #263, #262, #260, #251, #249.

`west-cascades`: Fixes [#159](https://github.com/OutdoorRD/west-cascades/issues/159), [#158](https://github.com/OutdoorRD/west-cascades/issues/158)

**Comment by jblehr** [comment]
Hey @davidye007, I'm confirming two things: I can run the dashboard locally on my computer (at least with firefox - I still run into [CORS issues with chrome](https://github.com/OutdoorRD/trails-viz/issues/186)), and that it looks awesome! You've built a ton of stuff on here and it's especially cool to see the new data that shows which sites are receiving chatbot data. 

I just had one main comment/question on the chatbot home locations process:
 
It seems like the chatbot home locations map and tree map functionality is identical to our generic "home locations" map and tree map functionality. And, my quick look at the code in several files shows significant duplication:
- `trails-viz-app/src/components/HomeLocations.vue` and `trails-viz-app/src/components/ChatbotHomeLocations.vue`
- `trails-viz-app/src/components/HomeLocationsMap.vue` and `trails-viz-app/src/components/ChatbotHomeLocationsMap.vue`

Further, within `trails-viz-api/trailsvizapi/repository/home_locations.py`, it also seems like there's a lot of similar code to process the chatbot home locations as the more general home locations workflow.

My main thought is: why can't we just use (more of) our existing home locations workflow and code to produce the home locations tree map and states/counties map? For example, for the `WestCascades` project, we have a `projects/WestCascades/homes_by_site_census.csv` file that is read in and displayed. The parallel here might be that https://github.com/OutdoorRD/vic-flaskbot/pull/241 (or a different script elsewhere) could just create a `trails-viz-data/chatbot/homes_by_site_census.csv`, then the rest of this workflow could largely use the existing code. 

I don't think this should hold up merging in this branch, but I'm curious to hear your thoughts and we can decide if it's worth creating an issue for this, etc.

**Review by jblehr — APPROVED** [review]

**Comment by davidye007** [comment]
@jblehr Thank you for testing out the dashboard and for your brilliant suggestions!

I really like your idea of merging the frontend components into a general `HomeLocations.vue` or `HomeLocationsMap.vue`, which accepts a parameter like `data_source='chatbot'` or `data_source='reveal'` to call the appropriate endpoint.

I also like your idea of using the existing `home locations` workflow for the API, and switching between the `chatbot` data and `reveal` data based on the parameter passed in.

I think implementing the approaches you suggested will significantly reduce code duplication, as well as make it easier to add new data sources in the future.

I'll create an issue for this!
